File: frontend/src/types/production.ts
export type ProductionStatus = 'registered' | 'approved' | 'started' | 'completed' | 'cancelled'

export type ProductionLineStatus = 'planned' | 'completed'

export type ProductionSourceType = 'manual' | 'excel'



export interface ProductionOrderLine {

  prd_order_line_id: number

  line_no: number

  process_no: number

  process_nm: string

  output_item_id: number | null

  output_item_cd: string | null

  output_item_nm: string | null

  planned_qty: string | number | null

  rm_location_id: number

  rm_location_cd: string

  wip_location_id: number

  wip_location_cd: string

  status: ProductionLineStatus

  actual_qty: string | number | null

  completed_at: string | null

}



export interface ProductionOrderInput {

  prd_order_input_id: number

  line_no: number

  level: number

  itemtyp_nm: string

  item_id: number

  item_cd: string

  item_nm: string

  from_location_id: number | null

  from_location_cd: string | null

  from_location_nm: string | null

  req_qty: string | number

  consume_qty: string | number

  lot: string | null

}



export interface ProductionOrderOutput {

  prd_order_output_id: number

  prd_order_line_id: number | null

  line_no: number

  item_id: number

  item_cd: string

  item_nm: string

  output_qty: string | number

  location_id: number

  location_cd: string

  location_nm: string

  lot: string

}



export interface ProductionOrderListItem {

  production_order_id: number

  status: ProductionStatus

  production_date: string

  reference_no: string | null

  source_type: ProductionSourceType

  parent_item_id: number

  parent_item_cd: string

  parent_item_nm: string

  planned_qty: string | number

  actual_qty: string | number | null

  lot: string

  line_count: number

  completed_line_count: number

  created_at: string | null

  approved_at: string | null

  cancelled_at: string | null

}



export interface ProductionOrderDetail extends ProductionOrderListItem {

  notes: string | null

  updated_at: string | null

  lines: ProductionOrderLine[]

  inputs: ProductionOrderInput[]

  outputs: ProductionOrderOutput[]

}



export interface ProductionOrderCreatePayload {

  production_date: string

  reference_no?: string | null

  parent_item_id: number

  planned_qty: number

  lot: string

  notes?: string | null

}



export interface ProductionOrderLineWritePayload {
  prd_order_line_id?: number
  line_no?: number
  rm_location_id: number
  wip_location_id: number
  output_item_id: number
  planned_qty: number
  actual_qty?: number | null
}

export interface ProductionOrderInputWritePayload {
  prd_order_input_id?: number
  item_id: number
  from_location_id: number
  req_qty: number
  consume_qty: number
  lot?: string | null
  line_no?: number
}



export interface ProductionOrderUpdatePayload {

  production_date?: string

  reference_no?: string | null

  planned_qty?: number

  actual_qty?: number

  lot?: string

  notes?: string | null

  status?: ProductionStatus

  lines?: ProductionOrderLineWritePayload[]

  inputs?: ProductionOrderInputWritePayload[]

}

File: backend/app/schemas/production.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

ProductionStatus = Literal["registered", "approved", "completed"]
ProductionLineStatus = Literal["planned", "completed"]
ProductionSourceType = Literal["manual", "excel"]


class ProductionOrderLineWrite(BaseModel):
    prd_order_line_id: int | None = None
    line_no: int | None = Field(default=None, gt=0)
    rm_location_id: int = Field(gt=0)
    wip_location_id: int = Field(gt=0)
    output_item_id: int = Field(gt=0)
    planned_qty: Decimal = Field(gt=0)
    actual_qty: Decimal | None = Field(default=None, gt=0)


class ProductionOrderInputWrite(BaseModel):
    prd_order_input_id: int | None = None
    item_id: int = Field(gt=0)
    from_location_id: int = Field(gt=0)
    req_qty: Decimal = Field(gt=0)
    consume_qty: Decimal = Field(gt=0)
    lot: str | None = Field(default=None, max_length=50)
    line_no: int | None = Field(default=None, gt=0)


class ProductionOrderCreate(BaseModel):
    production_date: date
    reference_no: str | None = Field(default=None, max_length=100)
    parent_item_id: int = Field(gt=0)
    planned_qty: Decimal = Field(gt=0)
    lot: str = Field(min_length=1, max_length=50)
    notes: str | None = None
    source_type: ProductionSourceType = "manual"


class ProductionOrderUpdate(BaseModel):
    production_date: date | None = None
    reference_no: str | None = Field(default=None, max_length=100)
    parent_item_id: int | None = Field(default=None, gt=0)
    planned_qty: Decimal | None = Field(default=None, gt=0)
    actual_qty: Decimal | None = Field(default=None, gt=0)
    lot: str | None = Field(default=None, min_length=1, max_length=50)
    notes: str | None = None
    status: ProductionStatus | None = None
    lines: list[ProductionOrderLineWrite] | None = None
    inputs: list[ProductionOrderInputWrite] | None = None


class ProductionOrderRecalculateIn(BaseModel):
    basis_qty: Decimal = Field(gt=0)


class ProductionOrderCompleteLineIn(BaseModel):
    actual_qty: Decimal = Field(gt=0)


class ProductionOrderInputRead(BaseModel):
    prd_order_input_id: int
    line_no: int
    level: int
    itemtyp_nm: str
    item_id: int
    item_cd: str
    item_nm: str
    from_location_id: int | None
    from_location_cd: str | None
    from_location_nm: str | None
    req_qty: Decimal
    consume_qty: Decimal
    lot: str | None


class ProductionOrderLineRead(BaseModel):
    prd_order_line_id: int
    line_no: int
    process_no: int
    process_nm: str
    output_item_id: int | None
    output_item_cd: str | None
    output_item_nm: str | None
    planned_qty: Decimal | None
    rm_location_id: int
    rm_location_cd: str
    wip_location_id: int
    wip_location_cd: str
    status: ProductionLineStatus
    actual_qty: Decimal | None
    completed_at: datetime | None


class ProductionOrderOutputRead(BaseModel):
    prd_order_output_id: int
    prd_order_line_id: int | None
    line_no: int
    item_id: int
    item_cd: str
    item_nm: str
    output_qty: Decimal
    location_id: int
    location_cd: str
    location_nm: str
    lot: str


class ProductionOrderListItem(BaseModel):
    production_order_id: int
    status: ProductionStatus
    production_date: date
    reference_no: str | None
    source_type: ProductionSourceType
    parent_item_id: int
    parent_item_cd: str
    parent_item_nm: str
    planned_qty: Decimal
    actual_qty: Decimal | None
    lot: str
    line_count: int
    completed_line_count: int
    created_at: datetime | None
    approved_at: datetime | None
    cancelled_at: datetime | None


class ProductionOrderRead(ProductionOrderListItem):
    notes: str | None
    updated_at: datetime | None
    lines: list[ProductionOrderLineRead]
    inputs: list[ProductionOrderInputRead]
    outputs: list[ProductionOrderOutputRead]


class ProductionExcelImportPreviewRow(BaseModel):
    excel_row: int
    action: Literal["insert", "update"]
    production_order_id: int | None = None
    production_date: date
    reference_no: str | None
    parent_item_id: int
    parent_item_cd: str
    parent_item_nm: str
    planned_qty: Decimal
    lot: str


class ProductionExcelImportResult(BaseModel):
    """Parsed Excel rows for grid merge; persist on client Update (no DB write here)."""

    rows: list[ProductionExcelImportPreviewRow] = []
    errors: list[str] = []


